components/raft_log_engine/src/engine.rs

```diff
@@ -89,74 +97,136 @@ impl<W: Seek + Write> Write for ManagedWriter<W> {
     }
 }
 
-struct ManagedFileBuilder {
+impl WriteExt for ManagedWriter {
+    fn truncate(&self, offset: usize) -> IoResult<()> {
+        match self.inner.as_ref() {
+            Either::Left(writer) => writer.truncate(offset),
+            Either::Right(writer) => writer.inner().truncate(offset),
+        }
+    }
+
+    fn sync(&self) -> IoResult<()> {
+        match self.inner.as_ref() {
+            Either::Left(writer) => writer.sync(),
+            Either::Right(writer) => writer.inner().sync(),
+        }
+    }
+
+    fn allocate(&self, offset: usize, size: usize) -> IoResult<()> {
+        match self.inner.as_ref() {
+            Either::Left(writer) => writer.allocate(offset, size),
+            Either::Right(writer) => writer.inner().allocate(offset, size),
+        }
+    }
+}
+
+struct ManagedFileSystem {
+    base_level_file_system: DefaultFileSystem,
     key_manager: Option<Arc<DataKeyManager>>,
     rate_limiter: Option<Arc<IORateLimiter>>,
 }
 
-impl ManagedFileBuilder {
+impl ManagedFileSystem {
     fn new(
         key_manager: Option<Arc<DataKeyManager>>,
         rate_limiter: Option<Arc<IORateLimiter>>,
     ) -> Self {
         Self {
+            base_level_file_system: DefaultFileSystem,
             key_manager,
             rate_limiter,
         }
     }
 }
 
-impl FileBuilder for ManagedFileBuilder {
-    type Reader<R: Seek + Read + Send> = ManagedReader<R>;
-    type Writer<W: Seek + Write + Send> = ManagedWriter<W>;
+struct ManagedHandle {
+    path: PathBuf,
+    base: Arc<<DefaultFileSystem as FileSystem>::Handle>,
```

Comment:
Is the `Arc` necessary?

Reply:
Yes. it will also be held by the base reader or writer
``` rust
fn new_reader(&self, handle: Arc<Self::Handle>) -> IoResult<Self::Reader> {
    Ok(LogFile::new(handle))
}
```